kotlinx-coroutines-core/common/src/CoroutineDispatcher.kt

```diff
@@ -65,37 +65,76 @@ public abstract class CoroutineDispatcher :
+     * It does not constitute a resource -- it is a _view_ of the underlying dispatcher that can be thrown away
+     * and is not required to be closed.
      *
      * ### Example of usage
      * ```
-     * private val backgroundDispatcher = newFixedThreadPoolContext(4, "App Background")
+     * // Background dispatcher for the application
+     * val dispatcher = newFixedThreadPoolContext(4, "App Background")
      * // At most 2 threads will be processing images as it is really slow and CPU-intensive
-     * private val imageProcessingDispatcher = backgroundDispatcher.limitedParallelism(2)
+     * val imageProcessingDispatcher = dispatcher.limitedParallelism(2)
      * // At most 3 threads will be processing JSON to avoid image processing starvation
-     * private val jsonProcessingDispatcher = backgroundDispatcher.limitedParallelism(3)
+     * val jsonProcessingDispatcher = dispatcher.limitedParallelism(3)
      * // At most 1 thread will be doing IO
-     * private val fileWriterDispatcher = backgroundDispatcher.limitedParallelism(1)
+     * val fileWriterDispatcher = dispatcher.limitedParallelism(1)
      * ```
      * Note how in this example the application has an executor with 4 threads, but the total sum of all limits
-     * is 6. Still, at most 4 coroutines can be executed simultaneously as each view limits only its own parallelism.
+     * is 6. Still, at most 4 coroutines can be executed simultaneously as each view limits only its own parallelism,
+     * and at most 4 threads can exist in the system.
      *
      * Note that this example was structured in such a way that it illustrates the parallelism guarantees.
      * In practice, it is usually better to use [Dispatchers.IO] or [Dispatchers.Default] instead of creating a
-     * `backgroundDispatcher`. It is both possible and advised to call `limitedParallelism` on them.
+     * `backgroundDispatcher`.
+     *
+     * ### `limitedParallelism(1)` pattern
+     *
+     * One of the common patterns is confining the execution of specific tasks to a sequential execution in background
+     * with `limitedParallelism(1)` invocation.
+     * For that purpose, the implementation guarantees that tasks are executed sequentially and that a happens-before relation
+     * is established between them:
+     *
+     * ```
+     * val confined = Dispatchers.Default.limitedParallelism(1)
+     * var counter = 0
+     *
+     * // Invoked from arbitrary coroutines
+     * launch(confined) {
+     *     // This increment is sequential and race-free
+     *     ++counter
+     * }
+     * ```
+     * Note that there is no guarantee that the underlying system thread will always be the same.
+     *
+     * ### Dispatchers.IO
+     *
+     * `Dispatcher.IO` is considered _elastic_ for the purposes of limited parallelism -- the sum of
+     * views is not restricted by the capacity of `Dispatchers.IO`.
+     * See `Dispatchers.IO` documentation for more details.
+     *
```

Comment:
I'm not sure this section belongs here, but this idea *is* missing from one other place: `newFixedThreadPoolContext`.

Who cares if `Dispatchers.IO` is elastic? Those who are afraid to hit the ceiling. In my view, those are:

* People who want to use `limitedParallelism` on `Dispatchers.IO`. They are covered already by the `Dispatchers.IO` documentation, as they will need to look at that documentation to see the if the default number of threads is enough for them.
* People who believe `Dispatchers.IO` isn't suitable for them, because they already know that it doesn't have a lot of threads. What code do these people have? It's probably something like this:
```kotlin
val dispatcher = newFixedThreadPoolContext(nThreads = 100, name = "")
```
We have a recommendation to replace these things with `limitedParallelism`, but we could certainly elaborate on the exact usage pattern and why it's safe.

WDYT?

Reply:
>Who cares if Dispatchers.IO is elastic?

I would say that for users who are looking for `limitedParallelism`, it's worth explicitly mentioning "hey, here is the common pattern that you might want to be aware of. It's briefly mentioned, see Dispatchers.IO" which seems a reasonable thing to do.

W.r.t. `newFixedThreadPoolContext` -- great idea, I'm on it

Reply:
I tried to be brief there, PTAL